### PR TITLE
docs: Batch Processing API doc

### PR DESCRIPTION
**Pull request by szewaiyuen6**
## Description
Add documentation for the new torch batch processing API

## Test Plan
- make live to ensure docs build
- Click through the docs to ensure it looks as expected

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

**Review by tara-det-ai — COMMENTED** [review]
i'll make some suggested edits via a commit to this PR

File: docs/reference/batch-processing/api-torch-batch-process-reference.rst [hunk 1]
```diff
@@ -0,0 +1,51 @@
+:orphan:
+
+.. _torch-batch-process-api-ref:
+
+#######################################
+ ``Torch Batch Process`` API Reference
+#######################################
+
```
Comment (tara-det-ai) [inline docs/reference/batch-processing/api-torch-batch-process-reference.rst:8]: something about the title does not look consistent with the other API ref page titles

should the title be more like this?

#######################################
 ``pytorch.torch_batch_process`` API Reference
#######################################

File: docs/reference/batch-processing/api-torch-batch-process-reference.rst [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
++============================================+
+| :ref:`torch-batch-processing-ug`           |
++--------------------------------------------+
+
+.. caution::
+
+   This is an experimental API and may change at any time.
+
+The main arguments to ``torch_batch_process`` is processor class and dataset.
+
```
Comment (tara-det-ai) [inline docs/reference/batch-processing/api-torch-batch-process-reference.rst:23]: The main arguments....are....
or
The main argument...is...

File: docs/model-dev-guide/batch-processing/batch-process-api-ug.rst [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+   :description: Learn how to use the Torch Batch Processing API.
+
+In this guide, you'll learn about the :ref:`Torch Batch Process API <torch-batch-process-api-ref>`
+and how to perform batch inference (also known as offline inference).
+
++---------------------------------------------------------------------+
+| Visit the API reference                                             |
++=====================================================================+
+| :ref:`torch-batch-process-api-ref`                                  |
++---------------------------------------------------------------------+
```
Comment (tara-det-ai) [inline docs/model-dev-guide/batch-processing/batch-process-api-ug.rst:17]: i'm guessing this might need to be a link to the class header and not the title of the page

we would have to create an anchor for it 

e.g., ".. _pytorch-batch-process:"

Then change the :ref: to 

:ref:`pytorch-batch-process`

File: docs/model-dev-guide/batch-processing/batch-process-api-ug.rst [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+The Torch Batch Processing API takes in (1) a dataset and (2) a user-defined processor class and
+runs distributed data processing.
+
+This API automatically handles the following for you:
+
+-  shard a dataset by number of workers available
+-  apply user-defined logic to each batch of data
+-  handle synchronization between workers
+-  track job progress to enable preemption and resumption of trial
+
```
Comment (tara-det-ai) [inline docs/model-dev-guide/batch-processing/batch-process-api-ug.rst:36]: This API allows you to accomplish the following tasks:

or

This API automatically handles the following for you:

-  shards a dataset by number of workers available
-  applies user-defined logic to each batch of data
-  handles synchronization between workers
-  tracks job progress to enable preemption and resumption of trial

File: docs/model-dev-guide/batch-processing/batch-process-api-ug.rst [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+This is a flexible API that can be used for many different tasks, including batch (offline)
+inference.
+
+*******
+ Usage
+*******
+
+The main arguments to :meth:`~determined.pytorch.experimental.torch_batch_process` are processor
+class and dataset.
```
Comment (tara-det-ai) [inline docs/model-dev-guide/batch-processing/batch-process-api-ug.rst:45]: The main arguments to :meth:`~determined.pytorch.experimental.torch_batch_process` are ``batch_processor_cls`` and ``dataset``.

File: docs/model-dev-guide/batch-processing/batch-process-api-ug.rst [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+           self.reducer = context.wrap_reducer(reducer=AccuracyMetricReducer(), name="accuracy")
+
+******************************************
+ How To Perform Batch (Offline) Inference
+******************************************
+
+In this section, we'll learn how to perform batch inference using the Torch Batch Processing API.
+
+Step 1: Define a InferenceProcessor
+===================================
```
Comment (tara-det-ai) [inline docs/model-dev-guide/batch-processing/batch-process-api-ug.rst:89]: Step 1: Define an InferenceProcessor

File: docs/model-dev-guide/batch-processing/batch-process-api-ug.rst [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+:meth:`~determined.pytorch.experimental.TorchBatchProcessor.__init__` function of the
+InferenceProcessor. You should implement
+:meth:`~determined.pytorch.experimental.TorchBatchProcessor.process_batch` function with inference
+logic.
+
+You can optionally implement
+:meth:`~determined.pytorch.experimental.TorchBatchProcessor.on_checkpoint_start` and
+:meth:`~determined.pytorch.experimental.TorchBatchProcessor.on_finish` function to be run before
+every checkpoint and after all the data are processed respectively.
+
```
Comment (tara-det-ai) [inline docs/model-dev-guide/batch-processing/batch-process-api-ug.rst:101]: You can optionally implement
:meth:`~determined.pytorch.experimental.TorchBatchProcessor.on_checkpoint_start` and
:meth:`~determined.pytorch.experimental.TorchBatchProcessor.on_finish` to be run before
every checkpoint and after all the data has been processed, respectively.

File: docs/model-dev-guide/batch-processing/batch-process-api-ug.rst [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+           """
+           if len(self.output) == 0:
+               return
+           file_name = f"prediction_output_{self.last_index}"
+           with self.context.upload_path() as path:
+               file_path = pathlib.Path(path, file_name)
+               torch.save(self.output, file_path)
+
+           self.output = []
+
```
Comment (tara-det-ai) [inline docs/model-dev-guide/batch-processing/batch-process-api-ug.rst:139]: If this is straight from the code in the repo, then we can use "literalinclude", e.g.,

.. literalinclude:: ../../../directoryl/filename.py
   :language: python
   :start-after: # Define custom processor class
   :end-after: # self.output = []
   :dedent:

even better if we add the Docs snippet comments to the code, e.g., # Docs snippet start: custom processor class, # Docs snippet end: custom processor class

Comment (szewaiyuen6) [reply docs/model-dev-guide/batch-processing/batch-process-api-ug.rst:139]: Discussed offline, leave as is for now.

**Review by tara-det-ai — APPROVED** [review]
LGTM

File: docs/model-dev-guide/batch-processing/batch-process-api-ug.rst (hunk 3 above)
Comment (tara-det-ai) [reply docs/model-dev-guide/batch-processing/batch-process-api-ug.rst:17]: we decided to link to the title of the page rather than a heading on the page.
